Add support for `azurerm_virtual_desktop_scaling_plan`

alexwilcox9 opened the pull request:
Adding support for AVD Scaling Plans

I've still got tests and docs to write but would appreciate if someone could have a quick read through of this and check they're happy with my implementation (especially around naming, I've followed the ARM template somewhat verbatim)

amccullough84 commented:
Good job I looked as I was about to make a pull request myself! I've already implemented acceptance tests so could pull into your branch?

Implementation and naming is effectively identical to mine, so minor tweaks to align

alexwilcox9 commented:
@amccullough84 that would be great, I did wonder what the odds were of someone else looking at this at the same time!

What's the best way to do this, I'm fairly new. I had a go at setting your branch as a remote and merging it with mine but it looked like it was pulling across a lot off extra stuff like the registration info resource from the other PR
Have I got the right branch?
https://github.com/amccullough84/terraform-provider-azurerm/tree/dv-scaling

amccullough84 commented:
@alexwilcox9 Yeah, mine had a few other bits and had dependencies on the other pull as I'd aligned the API version in all resources to 2021-09-03 - I'll add that as a separate one. I'm just cleaning up the acceptance tests to fully align with yours - had to  make a couple of tweaks as you need the role definition for autoscaling setup for the hostpool_reference to work.

I should (in theory) be able to send a pull request to your branch - just debugging an issue with role assignments in the test and will fire it over

amccullough84 commented:
@alexwilcox9 sent a pull request to your branch with acceptance tests, docs and a couple of minor tweaks in the resource (to fix test issues).

Acceptance tests are all clear now:

TF_ACC=1 go test -v ./internal/services/desktopvirtualization -run=TestAccVirtualDesktopScalingPlan -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccVirtualDesktopScalingPlan_basic
=== PAUSE TestAccVirtualDesktopScalingPlan_basic
=== RUN   TestAccVirtualDesktopScalingPlan_complete
=== PAUSE TestAccVirtualDesktopScalingPlan_complete
=== RUN   TestAccVirtualDesktopScalingPlan_update
=== PAUSE TestAccVirtualDesktopScalingPlan_update
=== RUN   TestAccVirtualDesktopScalingPlan_requiresImport
=== PAUSE TestAccVirtualDesktopScalingPlan_requiresImport
=== CONT  TestAccVirtualDesktopScalingPlan_basic
=== CONT  TestAccVirtualDesktopScalingPlan_requiresImport
=== CONT  TestAccVirtualDesktopScalingPlan_complete
=== CONT  TestAccVirtualDesktopScalingPlan_update
--- PASS: TestAccVirtualDesktopScalingPlan_basic (262.33s)
--- PASS: TestAccVirtualDesktopScalingPlan_requiresImport (442.57s)
--- PASS: TestAccVirtualDesktopScalingPlan_update (494.22s)
--- PASS: TestAccVirtualDesktopScalingPlan_complete (505.85s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/desktopvirtualization 506.020s

amccullough84 commented:
For transparency, I have deliberately left the hostpool_reference block out of the acceptance test configs. To add a hostpool_reference the "Windows Virtual Desktop" application needs a custom role creating and assigning to give permissions on the hostpool. Despite the fact the role is being created and assigned to application and scoped to the resource group in the acc test configuration the resource reports it does not have permission to access the hostpool. 

I've verified the exact same configuration works perfectly outside of the acceptance test, so can only presume it's some quirk of the acceptance test itself as the resource works perfectly for normal usage.

alexwilcox9 commented:
@amccullough84 thanks for the PR I've merged that in now after rebasing my stuff
I think you're right, the hostpool type needn't be specified when there's only one allowed value at the moment!

Re the reference block we shall see what the maintainers here think on that, I'll sort the linting test and mark this as ready
Thanks so much for your help on this!

catriona-m reviewed (changes requested):
Hi @alexwilcox9, thanks for this! Looks good so far, however I see there are some failing tests at the minute:

```
------- Stdout: -------
=== RUN   TestAccVirtualDesktopHostPool_update
=== PAUSE TestAccVirtualDesktopHostPool_update
=== CONT  TestAccVirtualDesktopHostPool_update
testcase.go:110: Step 2/3 error: After applying this test step, the plan was not empty.
stdout:
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
~ update in-place
Terraform will perform the following actions:
# azurerm_virtual_desktop_host_pool.test will be updated in-place
~ resource "azurerm_virtual_desktop_host_pool" "test" {
id                       = "/subscriptions/*******/resourceGroups/acctestRG-vdesktop-211117232024101875/providers/Microsoft.DesktopVirtualization/hostPools/acctestHPvr43e"
name                     = "acctestHPvr43e"
tags                     = {
"Purpose" = "Acceptance-Testing"
}
# (11 unchanged attributes hidden)
+ registration_info {
+ expiration_date = (known after apply)
}
}
Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccVirtualDesktopHostPool_update (179.74s)
FAIL
```

I also left a couple of minor suggestions inline.

amccullough84 commented:
> Hi @alexwilcox9, thanks for this! Looks good so far, however I see there are some failing tests at the minute:
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccVirtualDesktopHostPool_update
> === PAUSE TestAccVirtualDesktopHostPool_update
> === CONT  TestAccVirtualDesktopHostPool_update
> testcase.go:110: Step 2/3 error: After applying this test step, the plan was not empty.
> stdout:
> Terraform used the selected providers to generate the following execution
> plan. Resource actions are indicated with the following symbols:
> ~ update in-place
> Terraform will perform the following actions:
> # azurerm_virtual_desktop_host_pool.test will be updated in-place
> ~ resource "azurerm_virtual_desktop_host_pool" "test" {
> id                       = "/subscriptions/*******/resourceGroups/acctestRG-vdesktop-211117232024101875/providers/Microsoft.DesktopVirtualization/hostPools/acctestHPvr43e"
> name                     = "acctestHPvr43e"
> tags                     = {
> "Purpose" = "Acceptance-Testing"
> }
> # (11 unchanged attributes hidden)
> + registration_info {
> + expiration_date = (known after apply)
> }
> }
> Plan: 0 to add, 1 to change, 0 to destroy.
> --- FAIL: TestAccVirtualDesktopHostPool_update (179.74s)
> FAIL
> ```
> 
> I also left a couple of minor suggestions inline.

@catriona-m the failing test seems to be a regression outside of this PR - the acceptance test passes against the 2.85 provider but fails against the 2.86 provider.

There is a lifecycle block in the acceptance test to ignore diffs on the expiration date as it's a relative date being used so would cause a perpetual diff.

```
  lifecycle {
    ignore_changes = [
      registration_info[0].expiration_date,
    ]
  }
```
This now does not work in 2.86 unless the attribute reference is changed to:

```
  lifecycle {
    ignore_changes = [
      registration_info.expiration_date,
    ]
  }
```

We could update the acceptance test, but seems like there's a regression here?

alexwilcox9 commented:
HI @catriona-m I've added the requested changes now, good spot!

As @amccullough84 says I'm not sure why this is now failing, I have made the change suggested anyway and hopefully if you test again that will work but might be worth some more digging

alexwilcox9 commented:
So by changing 
```
  lifecycle {
    ignore_changes = [
      registration_info[0].expiration_date,
    ]
  }
```
to
```
  lifecycle {
    ignore_changes = [
      registration_info[0].expiration_date,
    ]
  }
```
I started to get the following error
```
=== CONT  TestAccVirtualDesktopHostPool_update
    testcase.go:110: Step 2/3 error: Error running pre-apply refresh: exit status 1
        
        Error: Invalid operation
        
          on terraform_plugin_test.tf line 31, in resource "azurerm_virtual_desktop_host_pool" "test":
          31:       registration_info.expiration_date,
        
        Block type "registration_info" is represented by a list of objects, so it
        must be indexed using a numeric key, like .registration_info[0].
    testing_new.go:70: Error running post-test destroy, there may be dangling resources: exit status 1
        
        Error: Invalid operation
        
          on terraform_plugin_test.tf line 31, in resource "azurerm_virtual_desktop_host_pool" "test":
          31:       registration_info.expiration_date,
        
        Block type "registration_info" is represented by a list of objects, so it
        must be indexed using a numeric key, like .registration_info[0].
```


Instead I've updated the `expiration_date` to exclude the time part of the timestamp and removed the `ignore_changes` block entirely

Is that suitable for the tests?

alexwilcox9 commented:
Test output:
```
$ make acctests SERVICE=desktopvirtualization TESTARGS='-run=TestAccVirtualDesktop' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/desktopvirtualization -run=TestAccVirtualDesktop -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccVirtualDesktopApplicationGroup_basic
=== PAUSE TestAccVirtualDesktopApplicationGroup_basic
=== RUN   TestAccVirtualDesktopApplicationGroup_complete
=== PAUSE TestAccVirtualDesktopApplicationGroup_complete
=== RUN   TestAccVirtualDesktopApplicationGroup_update
=== PAUSE TestAccVirtualDesktopApplicationGroup_update
=== RUN   TestAccVirtualDesktopApplicationGroup_requiresImport
=== PAUSE TestAccVirtualDesktopApplicationGroup_requiresImport
=== RUN   TestAccVirtualDesktopApplication_basic
=== PAUSE TestAccVirtualDesktopApplication_basic
=== RUN   TestAccVirtualDesktopApplication_complete
=== PAUSE TestAccVirtualDesktopApplication_complete
=== RUN   TestAccVirtualDesktopApplication_update
=== PAUSE TestAccVirtualDesktopApplication_update
=== RUN   TestAccVirtualDesktopApplication_requiresImport
=== PAUSE TestAccVirtualDesktopApplication_requiresImport
=== RUN   TestAccVirtualDesktopHostPool_basic
=== PAUSE TestAccVirtualDesktopHostPool_basic
=== RUN   TestAccVirtualDesktopHostPool_complete
=== PAUSE TestAccVirtualDesktopHostPool_complete
=== RUN   TestAccVirtualDesktopHostPool_update
=== PAUSE TestAccVirtualDesktopHostPool_update
=== RUN   TestAccVirtualDesktopHostPool_requiresImport
=== PAUSE TestAccVirtualDesktopHostPool_requiresImport
=== RUN   TestAccVirtualDesktopScalingPlan_basic
=== PAUSE TestAccVirtualDesktopScalingPlan_basic
=== RUN   TestAccVirtualDesktopScalingPlan_complete
=== PAUSE TestAccVirtualDesktopScalingPlan_complete
=== RUN   TestAccVirtualDesktopScalingPlan_update
=== PAUSE TestAccVirtualDesktopScalingPlan_update
=== RUN   TestAccVirtualDesktopScalingPlan_requiresImport
=== PAUSE TestAccVirtualDesktopScalingPlan_requiresImport
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport
=== CONT  TestAccVirtualDesktopApplicationGroup_basic
=== CONT  TestAccVirtualDesktopHostPool_requiresImport
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete
=== CONT  TestAccVirtualDesktopApplication_update
=== CONT  TestAccVirtualDesktopScalingPlan_basic
=== CONT  TestAccVirtualDesktopScalingPlan_complete
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport
=== CONT  TestAccVirtualDesktopHostPool_update
=== CONT  TestAccVirtualDesktopScalingPlan_requiresImport
=== CONT  TestAccVirtualDesktopScalingPlan_update
--- PASS: TestAccVirtualDesktopApplicationGroup_basic (83.92s)
=== CONT  TestAccVirtualDesktopHostPool_complete
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete (86.77s)
=== CONT  TestAccVirtualDesktopHostPool_basic
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic (87.17s)
=== CONT  TestAccVirtualDesktopApplication_requiresImport
--- PASS: TestAccVirtualDesktopHostPool_requiresImport (105.22s)
=== CONT  TestAccVirtualDesktopApplicationGroup_requiresImport
--- PASS: TestAccVirtualDesktopHostPool_complete (71.93s)
=== CONT  TestAccVirtualDesktopApplication_complete
--- PASS: TestAccVirtualDesktopHostPool_basic (69.69s)
=== CONT  TestAccVirtualDesktopApplication_basic
--- PASS: TestAccVirtualDesktopScalingPlan_basic (169.16s)
=== CONT  TestAccVirtualDesktopApplicationGroup_update
--- PASS: TestAccVirtualDesktopHostPool_update (169.35s)
=== CONT  TestAccVirtualDesktopApplicationGroup_complete
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport (169.39s)
--- PASS: TestAccVirtualDesktopApplication_update (180.43s)
--- PASS: TestAccVirtualDesktopApplication_requiresImport (97.95s)
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update (185.26s)
--- PASS: TestAccVirtualDesktopScalingPlan_requiresImport (190.70s)
--- PASS: TestAccVirtualDesktopApplicationGroup_requiresImport (95.90s)
--- PASS: TestAccVirtualDesktopApplication_basic (73.50s)
--- PASS: TestAccVirtualDesktopApplication_complete (75.01s)
--- PASS: TestAccVirtualDesktopApplicationGroup_complete (73.23s)
--- PASS: TestAccVirtualDesktopApplicationGroup_update (163.23s)
--- PASS: TestAccVirtualDesktopScalingPlan_complete (420.24s)
--- PASS: TestAccVirtualDesktopScalingPlan_update (584.86s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/desktopvirtualization 584.917s
```

amccullough84 commented:
> Is that suitable for the test?

I'd say yes as that should be deprecated once #14134 is merged.

tombuildsstuff reviewed (changes requested):
hey @alexwilcox9 

Thanks for this PR / making those changes - I've taken a look through and left some comments inline in addition to what @catriona-m has mentioned, but if we can fix those up then this should otherwise be good to go 👍 

Thanks!

alexwilcox9 commented on internal/services/desktopvirtualization/virtual_desktop_scaling_plan_resource.go:
```diff
@@ -0,0 +1,524 @@
+
+			"hostpool_reference": {
+				Type:     pluginsdk.TypeList,
+				Optional: true,
```
Hi @tombuildsstuff, thanks for the review
I've done some testing on this and it is possible to create a scaling plan and not assign it to any hostpool. A little pointless but possible

alexwilcox9 commented on internal/services/desktopvirtualization/virtual_desktop_scaling_plan_resource.go:
```diff
@@ -0,0 +1,524 @@
+					Schema: map[string]*pluginsdk.Schema{
+						"hostpool_id": {
+							Type:         pluginsdk.TypeString,
+							Required:     true,
```
Similarly I've been able to remove and add hostpools without creating new scaling plans each time

alexwilcox9 commented on website/docs/r/virtual_desktop_scaling_plan.html.markdown:
```diff
@@ -0,0 +1,246 @@
+The following arguments are supported:
+
+
+* `location` - (Required) The Azure Region where the Virtual Desktop Scaling Plan  should exist. Changing this forces a new Virtual Desktop Scaling Plan to be created. **Note**: During preview only EU and US regions are supported.
```
Yeah that's a good idea, change has been made

alexwilcox9 commented on website/docs/r/virtual_desktop_scaling_plan.html.markdown:
```diff
@@ -0,0 +1,246 @@
+
+A `hostpool_reference` block supports the following:
+
+* `hostpool_id` - (Required) The ID of the HostPool to assign the Scaling Plan to. Note: During preview only HostPool's in EU and US regions are supported.
```
Changed as above

alexwilcox9 commented on website/docs/r/virtual_desktop_scaling_plan.html.markdown:
```diff
@@ -0,0 +1,246 @@
+
+* `hostpool_id` - (Required) The ID of the HostPool to assign the Scaling Plan to. Note: During preview only HostPool's in EU and US regions are supported.
+
+* `scaling_plan_enabled` - (Required) (boolean) Whether the scaling plan is enabled or disabled for the HostPool.
```
Change made as suggested

alexwilcox9 commented on website/docs/r/virtual_desktop_scaling_plan.html.markdown:
```diff
@@ -0,0 +1,246 @@
+
+* `ramp_down_load_balancing_algorithm` - (Required) The load Balancing Algorithm to use during the Ramp-Down period. Possible values are `DepthFirst` and `BreadthFirst`.
+
+* `ramp_down_minimum_hosts_percent` - (Required) The minimum percentage of session host virtual machines that you would like to get to for ramp-down and off-peak hours. For example, if Minimum percentage of hosts is sepecified as 10% and total number of session hosts in your host pool is 10 , autoscale will ensure a minimum of 1 session host is available to take user connections..
```
Good spot thanks

alexwilcox9 commented on internal/services/desktopvirtualization/virtual_desktop_scaling_plan_resource.go:
```diff
@@ -0,0 +1,524 @@
+						"off_peak_start_time": {
+							Type:     pluginsdk.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringMatch(regexp.MustCompile(`^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$`),
```
Didn't even occur to me that was something you could do! I've done that now, thanks for the tip

alexwilcox9 replied:
@tombuildsstuff could you check how I've implemented this ask? Now getting linting errors telling me that `schema.SchemaValidateFunc is deprecated: please use SchemaValidateDiagFunc`
I can't find a `StringMatch` function that returns a `SchemaValidateDiagFunc` rather than a `SchemaValidateFunc`

Am I missing something or do I need to put something around `StringMatch` to make is a `SchemaValidateDiagFunc`

katbyte reviewed (approved):
Thanks @alexwilcox9 - LGTM 🌳
